tests/db_tests/test_queries.py

```diff
@@ -481,3 +484,53 @@ def test_project_if_no_outcomes(seeded_test_client, additional_test_data):
 
     test_df = pd.read_sql(test_query_proj.statement, con=db.engine.connect())
     assert list(test_df["project_name"])
+
+
+def test_database_transaction_wrapper_max_retries(mocker, test_client, caplog):
+    """
+    Test the behavior of 'databse_transaction' wrapper in case of IntegrityError.
+
+    This test simulates a scenario where a function raises an IntegrityError
+    and ensures that the 'database_transaction' wrapper retries the operation up to the
+    specified maximum number of times. It also checks whether the error is properly
+    logged after reaching the maximum number of retries.
+    """
+    max_retries = 3
+    sleep_duration = 0.1
+    error_type = exc.IntegrityError
+
+    mocked_function = mocker.Mock(side_effect=exc.IntegrityError("Mocked IntegrityError", None, None))
+    mocked_function.__name__ = "mocked_function"
+    decorated_function = database_transaction(max_retries, sleep_duration, error_type)(mocked_function)
+
+    with caplog.at_level(logging.ERROR) and patch("time.sleep") as patched_time_sleep:
+        with pytest.raises(error_type):
+            decorated_function()
+
+    assert mocked_function.call_count == max_retries
+    assert caplog.messages[-1] == "Number of max retries exceeded for function 'mocked_function'"
+    assert patched_time_sleep.call_count == 2
+
+
+def test_database_transaction_wrapper_successful_retry(mocker, test_client, caplog):
+    """
+    Test the behavior of 'database_transaction' wrapper succeeding after failing once.
+    """
+    max_retries = 3
+    sleep_duration = 0.1
+    error_type = exc.IntegrityError
+
+    def mocked_side_effect(*args, **kwargs):
```

Comment:
I think a simpler way to achieve this is to just pass the expected side_effects as a list (the param can take an iterable). If you enter None, then for that iteration it doesn't raise. Ie, remove this mocked_side_effect sub-function and just use the following when you define mocked_function:

```
    mocked_function = mocker.Mock(side_effect=[
        exc.IntegrityError("Mocked IntegrityError", None, None),
        None,
        None,
    ])
```

Reply:
I tested this on mine locally and it worked